entmax/root_finding.py

```diff
@@ -65,6 +62,56 @@ def forward(ctx, X, n_iter=50, ensure_sum_one=True):
 
         return p_m
 
+    @classmethod
+    def backward(cls, ctx, dY):
+        Y, = ctx.saved_tensors
+
+        gppr = torch.where(Y > 0, Y ** (2 - ctx.alpha), Y.new_zeros(1))
+
+        dX = dY * gppr
+        q = dX.sum(ctx.dim) / gppr.sum(ctx.dim)
+        q = q.unsqueeze(ctx.dim)
+        dX -= q * gppr
+
+        d_alpha = None
+        if ctx.needs_input_grad[1]:
+
+            # alpha gradient computation
+            # d_alpha = (partial_y / partial_alpha) * dY
+            # NOTE: ensure alpha is not close to 1
+            # since there is an indetermination
+            batch_size, _ = dY.shape
+
+            # shannon terms
+            S = torch.where(Y > 0, Y * torch.log(Y), Y.new_zeros(1))
+            # shannon entropy
+            ent = S.sum(ctx.dim).unsqueeze(-1)
+            Y_skewed = gppr / gppr.sum(ctx.dim).unsqueeze(-1)
+            d_alpha = dY * (Y - Y_skewed) / ((ctx.alpha - 1) ** 2)
+            d_alpha -= dY * (S - Y_skewed * ent) / (ctx.alpha - 1)
+            d_alpha = d_alpha.sum(dim=-1)
+
+        return dX, d_alpha, None, None
+
+
+# slightly more efficient special case for sparsemax
+class SparsemaxBisectFunction(EntmaxBisectFunction):
```

Comment:
@bpopeters here is where i saved a lot of space by sharing code using `@classmethods`.

Reply:
Looks great!